src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java

```diff
@@ -81,4 +89,29 @@ public Date deserialize(JsonElement jsonElement, Type type,
             }
         }
     }
+
+    private class StatsDeserializer implements JsonDeserializer<Stats> {
+
+        @Override
+        public Stats deserialize(JsonElement jsonElement, Type type,
+                JsonDeserializationContext jsonDeserializationContext)
+                throws JsonParseException {
+            try {
+                JsonObject stats = jsonElement.getAsJsonObject();
+                Applications app = gson.fromJson(stats.get("CherryPy Applications"), Applications.class);
+                HttpServer server = null;
+                for(Map.Entry<String, JsonElement> entry : stats.entrySet()){
+                    String key = entry.getKey();
+                    if(key.startsWith("CherryPy HTTPServer ")){
+                        server = gson.fromJson(entry.getValue(), HttpServer.class);
+                        break;
+                    }
+                }
+                return new Stats(app, server);
```

Comment:
*[I am being a bit more verbose here since similar discussions happen routinely, so I hope to set this topic straight once and for all]*

Deciding whether to return `null` from an API should always be driven by the following question, which I also encourage to apply here:

**Is `null` expected as a particular but possible case or does it mean that an error has happened somewhere?**

If `null` is just one output possibility and really indicates a missing value that is optional in its nature, but it signals no error per se, then you should return `null`, document it and have the caller deal with it. If caller's code ends up with a nasty `NullPointerException` because they forgot to handle the `null` case properly, that's their fault, they should fix it and you don't have to worry. One day Java 8 arrives and you will be able to do the same, with compiler assistance by using `Optional`.

If `null` in your output is not really reasonable, but rather tells that your code has encountered a problem, then you should not return it. As usual you have two possibilities: either handle the error in your method, if you have enough information to do so, or signal the problem to the caller by throwing an appropriate `Exception` and move the error handling burden one layer up.

@lucidd: in this particular case, can you tell if a server element is always expected to be returned from the CherryPy API? If it is always expected, then you should check that it is really there and throw `Exception` if it is not - CherryPy is returning incomplete data so there's a problem below your level, and whoever calls you should know as you can't possibly handle the issue in here.

Otherwise you should be able to answer this question: in what reasonable cases do we expect a server to be missing from CherryPy's API output?